### PR TITLE
feat(jsx): inline multi-return JSX helper functions as conditional IR

### PR DESCRIPTION
**Pull request by kfly8**
## Summary

- Local helper functions with multiple JSX return statements (if/else chains, switch/case, guard clauses) are now inlined at call sites as nested `IRConditional` nodes, instead of falling through to opaque dynamic expressions.
- Adds `jsxMultiReturnFunctions` map to `AnalyzerContext` with `extractMultiReturnJsxBranches` extraction in the analyzer, and `transformMultiReturnJsxFunctionCall` in `jsx-to-ir.ts` for IR-level conditional synthesis with parameter substitution.
- Supported patterns: `if/else`, `if/else if/else` chains, guard clauses (`if (!x) return null`), `switch/case` statements, and arrow function variants.

## Test plan

- [x] Existing 1636 compiler unit tests pass (0 regressions)
- [x] Existing 497 adapter conformance tests pass (0 regressions)
- [x] New analyzer test: multi-return functions stored in `jsxMultiReturnFunctions`
- [x] New IR test: if/else helper inlined as conditional IR
- [x] New IR test: if/else if/else chain produces nested conditionals
- [x] New IR test: guard clause (`return null`) handled correctly
- [x] New IR test: arrow function with multi-return is inlined
- [x] New IR test: parameter substitution works in both conditions and branches

https://claude.ai/code/session_01SEG1QwvsXVy6LoAFbYXLLS

---
_Generated by [Claude Code](https://claude.ai/code/session_01SEG1QwvsXVy6LoAFbYXLLS)_

### CODE REVIEW

File: packages/jsx/src/analyzer.ts [hunk 1]
```diff
@@ -1834,6 +1834,138 @@ function extractSingleJsxReturn(
...
+            jsxReturn: jsxReturn ?? null,
+          })
+        } else {
+          fallback = jsxReturn ?? null
+        }
+      }
+
+      if (branches.length === 0) return null
+      return { branches, fallback, switchDiscriminant: stmt.expression }
+    }
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/analyzer.ts:1928]: Fixed — mixed if+switch bodies now return `null` (rejected) when `branches.length > 0` before entering the switch path.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/analyzer.ts:1928]: Added three switch tests: basic switch/case inlining, switch with null default, and a guard that switch with call-expression discriminant is NOT inlined.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/jsx-to-ir.ts [hunk 2]
```diff
@@ -1446,6 +1446,108 @@ function transformJsxFunctionCall(
...
+
+      // Build condition text with param substitution
+      let conditionText: string
+      if (info.switchDiscriminant) {
+        const discText = substitutedGetJS(info.switchDiscriminant)
+        const caseText = substitutedGetJS(branch.condition)
+        conditionText = `${discText} === ${caseText}`
+      } else {
+        conditionText = substitutedGetJS(branch.condition)
+      }
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/jsx-to-ir.ts:1507]: Fixed — `extractMultiReturnJsxBranches` now rejects switch statements whose discriminant is not an identifier or property-access expression. Added a test verifying call-expression discriminants are not inlined.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/jsx-to-ir.ts [hunk 3]
```diff
@@ -1446,6 +1446,108 @@ function transformJsxFunctionCall(
...
+        effect: 'pure',
+        freeRefs: resolveFreeRefs(branch.condition, makeBindingEnv(ctx)),
+      }
+      const reactive = isReactiveExpression(conditionText, ctx, branch.condition)
+        || isReactiveOrigin(conditionOrigin)
+      const loopParamReactive = !reactive && referencesLoopParam(conditionText, ctx)
+      const callsReactive = exprCallsReactiveGetters(branch.condition, ctx)
+      const hasCalls = exprHasFunctionCalls(branch.condition)
+      const needsSlot = reactive || loopParamReactive || callsReactive || hasCalls
+      const slotId = needsSlot ? generateSlotId(ctx) : null
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/jsx-to-ir.ts:1531]: This is a pre-existing limitation shared with the single-return `transformJsxFunctionCall` (line 1416–1422), which uses the same text-level `getJS` override and original-AST origin. The main reactive flag IS computed from the substituted string (`isReactiveExpression(conditionText, ...)`) so signal calls introduced via substitution are detected for the `reactive` flag. The `freeRefs` from the original AST are less precise but match the existing pattern. Moving to AST-level substitution is a broader refactor that would benefit both paths — out of scope for this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/jsx-to-ir.ts [hunk 4]
```diff
@@ -1446,6 +1446,108 @@ function transformJsxFunctionCall(
...
+  const baseGetJS = ctx.analyzer.getJS.bind(ctx.analyzer)
+  const originalCtxGetJS = ctx.getJS
+  const originalAnalyzerGetJS = ctx.analyzer.getJS
+
+  const substitutedGetJS = (node: ts.Node) => {
+    let text = baseGetJS(node)
+    for (const [paramName, argExpr] of substitutions) {
+      text = text.replace(new RegExp(`\\b${paramName}\\b`, 'g'), argExpr)
+    }
+    return text
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/jsx-to-ir.ts:1473]: This is the same `\b`-based regex pattern used by the existing `transformJsxFunctionCall` (line 1419). Both paths share this limitation. A broader refactor to AST-level identifier substitution would fix both — tracked as a future improvement, out of scope here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/jsx-to-ir.ts [hunk 5]
```diff
@@ -1446,6 +1446,108 @@ function transformJsxFunctionCall(
...
+    for (let i = info.branches.length - 1; i >= 0; i--) {
+      const branch = info.branches[i]
+
+      // Build condition text with param substitution
+      let conditionText: string
+      if (info.switchDiscriminant) {
+        const discText = substitutedGetJS(info.switchDiscriminant)
+        const caseText = substitutedGetJS(branch.condition)
+        conditionText = `${discText} === ${caseText}`
+      } else {
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/jsx-to-ir.ts:1505]: Fixed — for switch-sourced conditions, `resolveFreeRefs`, `exprCallsReactiveGetters`, and `exprHasFunctionCalls` now merge results from both the discriminant and case expression AST nodes. `templateCondition` is also built by rewriting each part separately to cover prop references in the discriminant.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/analyzer.ts [hunk 6]
```diff
@@ -1834,6 +1834,150 @@ function extractSingleJsxReturn(
...
+    if (ts.isIfStatement(stmt)) {
+      // Walk this if/else if/else chain
+      let current: ts.Statement = stmt
+      while (ts.isIfStatement(current)) {
+        const ifStmt = current
+        const jsxReturn = findJsxReturnInBlock(ifStmt.thenStatement)
+        const nullReturn = findNullReturnInBlock(ifStmt.thenStatement)
+        if (!jsxReturn && !nullReturn) return null
+
+        // Include both JSX and null-returning branches (guard clauses)
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/analyzer.ts]: Fixed — added `isDirectReturnBlock` guard that rejects if-blocks containing nested control flow (nested if/switch/for/while/try) or multiple return statements. Each then/else arm must be a direct `return` or a block with exactly one return.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/analyzer.ts [hunk 7]
```diff
@@ -1834,6 +1834,150 @@ function extractSingleJsxReturn(
...
+      } else if (expr.kind === ts.SyntaxKind.NullKeyword) {
+        // fallback stays null
+      } else {
+        return null
+      }
+      continue
+    }
+
+    // Skip variable declarations (they can appear before if/switch)
+    if (ts.isVariableStatement(stmt)) continue
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/analyzer.ts]: Fixed — `VariableStatement` now causes `return null` (rejecting the body) instead of `continue`. Function bodies with local variable declarations are not eligible for multi-return inlining.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/__tests__/jsx-function-inlining.test.ts [hunk 8]
```diff
@@ -422,4 +427,217 @@ describe('JSX function inlining (#569)', () => {
       expect(template!.content).toContain('table')
     })
   })
+
+  describe('multi-return JSX function inlining', () => {
+    test('if/else helper is inlined as conditional IR', () => {
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/__tests__/jsx-function-inlining.test.ts:432]: Added IR structure assertion (verifies a `conditional` node exists in the IR tree) and `clientJs` assertion (verifies the helper function is not emitted verbatim) to the first multi-return test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/analyzer.ts [hunk 9]
```diff
@@ -1834,6 +1834,185 @@ function extractSingleJsxReturn(
...
+            current = ifStmt.elseStatement
+            continue
+          }
+          if (!isDirectReturnBlock(ifStmt.elseStatement)) return null
+          const elseJsx = findJsxReturnInBlock(ifStmt.elseStatement)
+          if (elseJsx) {
+            fallback = elseJsx
+          } else if (!findNullReturnInBlock(ifStmt.elseStatement)) {
+            return null
+          }
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/analyzer.ts:1883]: Fixed — after processing a final `else` block (not `else if`), the extractor now returns `{ branches, fallback }` immediately, preventing any subsequent `return` from overwriting the fallback.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/analyzer.ts [hunk 10]
```diff
@@ -1834,6 +1834,185 @@ function extractSingleJsxReturn(
...
+          branches.push({
+            condition: clause.expression,
+            jsxReturn: jsxReturn ?? null,
+          })
+        } else {
+          fallback = jsxReturn ?? null
+        }
+      }
+
+      if (branches.length === 0) return null
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/analyzer.ts:1926]: Fixed — switches without an explicit `default` clause are now rejected (`return null`). Added a test verifying this.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/jsx/src/__tests__/jsx-function-inlining.test.ts [hunk 11]
```diff
@@ -422,4 +427,236 @@ describe('JSX function inlining (#569)', () => {
...
+      const source = `
+        'use client'
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/__tests__/jsx-function-inlining.test.ts:434]: Added two edge case tests: (1) switch without default is rejected, (2) if/else fallback is preserved and not overwritten by trailing statements.

---
_Generated by [Claude Code](https://claude.ai/code)_
